### DISCUSSION TITLE
Excel VBA Help?

### DISCUSSION THREAD

**Discussion by Cart-2004** [community/community#134856] [Programming Help]
### Body

Hi, I have written a code in Excel VBA that requires searching for certain numbers (item number) in combination with another values (sizes). Now the found value have to be extracted from the "stock" in a chronological way (table with article number vertically and sizes horizontally). So check if the value in the table is above 0 then it is copied to certain cell (shopping cart). If the value in the table is less than or equal to 0 the value is put into another cell (reserve list). Now the values are put in the reserve list no matter which value is in the "stock table". How can I solve this? 

The code:
Sub FindNumbersAndSizes()
    Dim ws As Worksheet
    Dim rng As Range
    Dim cell As Range
    Dim numberFound As String
    Dim sizeFound As String
    Dim numbers As Variant
    Dim sizes As Variant
    Dim extraSizes As Variant
    Dim combinedSizes As Variant
    Dim i As Integer
    Dim j As Integer
    Dim nameFound As String
    Dim words As Variant
    Dim word As Variant
    Dim cleanedWord As String
    Dim combined As String
    Dim matchFound As Boolean
    Dim isNumberAbove250 As Boolean
    Dim stockSheet As Worksheet
    Dim stockRange As Range
    Dim stockCell As Range
    Dim stockValue As Long
    
    ' Set the worksheet (change "Sheet1" to the name of your sheet)
    Set ws = ThisWorkbook.Sheets("Liveshow")
    
    ' Set the stock sheet (change "StockSheet" to the name of your stock sheet)
    Set stockSheet = ThisWorkbook.Sheets("StockSheet")
    
    ' Set the range to the column you want to search (e.g., column C)
    Set rng = ws.Range("C1:C" & ws.Cells(ws.Rows.Count, "C").End(xlUp).Row)
    
    ' Define the numbers and sizes to search for
    numbers = Array(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, _
                    21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, _
                    50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, _
                    71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 86, 87, 88, 89, 90, _
                    91, 92, 93, 94, 96, 97, 98, 99, 100, _
                    101, 102, 103, 104, 106, 107, 108, 109, 110, _
                    111, 112, 113, 114, 115, 116, 117, 118, 119, 120, _
                    121, 122, 123, 124, 125, 126, 127, 128, 129, 130, _
                    131, 132, 133, 134, 135, 136, 137, 138, 139, 140, _
                    141, 142, 143, 144, 145, 146, 147, 148, 149, 150, _
                    151, 152, 153, 154, 155, 156, 157, 158, 159, 160, _
                    161, 162, 163, 164, 165, 166, 167, 168, 169, 170, _
                    171, 172, 173, 174, 175, 176, 177, 178, 179, 180, _
                    181, 182, 183, 184, 185, 186, 187, 188, 189, 190, _
                    191, 192, 193, 194, 195, 196, 197, 198, 199, 200, _
                    201, 202, 203, 204, 205, 206, 207, 208, 209, 210, _
                    211, 212, 213, 214, 215, 216, 217, 218, 219, 220, _
                    221, 222, 223, 224, 225, 226, 227, 228, 229, 230, _
                    231, 232, 233, 234, 235, 236, 237, 238, 239, 240, _
                    241, 242, 243, 244, 245, 246, 247, 248, 249, 250)
    
    ' Define size mapping
    sizes = Array("34", "xs", "extra small", _
                  "36", "s", "small", _
                  "38", "m", "medium", _
                  "40", "l", "large", _
                  "42", "xl", "extra large", _
                  "44", "xxl", "2xl", _
                  "46", "xxxl", "3xl")
    
    ' Define the combined sizes to search for
    combinedSizes = Array("s/m", "m/l", "l/xl")
    
    ' Define extra sizes to search for
    extraSizes = Array("85", "95", "105")
    
    ' Loop through each cell in the range
    For Each cell In rng
        numberFound = ""
        sizeFound = ""
        nameFound = ""
        matchFound = False
        isNumberAbove250 = False
        
        ' Get the name from column B
        nameFound = cell.Offset(0, -1).Value
        
        ' Split the cell value into words
        words = Split(cell.Value, " ")
        
        ' Check each word for numbers and sizes
        For Each word In words
            ' Remove periods and dashes from the word
            cleanedWord = Replace(word, ".", "")
            cleanedWord = Replace(cleanedWord, "-", "")
            
            ' Check for numbers
            For i = LBound(numbers) To UBound(numbers)
                If cleanedWord = CStr(numbers(i)) Then
                    numberFound = CStr(numbers(i))
                    Exit For
                End If
            Next i
            
            ' Check for sizes
            For i = LBound(sizes) To UBound(sizes) Step 3
                If LCase(cleanedWord) = LCase(CStr(sizes(i))) Or _
                   LCase(cleanedWord) = LCase(CStr(sizes(i + 1))) Or _
                   LCase(cleanedWord) = LCase(CStr(sizes(i + 2))) Then
                    sizeFound = CStr(sizes(i))
                    Exit For
                End If
            Next i
            
            ' Check for extra sizes
            If sizeFound = "" Then
                For i = LBound(extraSizes) To UBound(extraSizes)
                    If LCase(cleanedWord) = LCase(extraSizes(i)) Then
                        sizeFound = CStr(extraSizes(i))
                        Exit For
                    End If
                Next i
            End If
            
            ' Check for combined numbers and sizes
            If numberFound = "" Or sizeFound = "" Then
                For i = LBound(numbers) To UBound(numbers)
                    For j = LBound(sizes) To UBound(sizes)
                        combined = CStr(numbers(i)) & CStr(sizes(j))
                        If InStr(LCase(cleanedWord), LCase(combined)) > 0 Then
                            numberFound = CStr(numbers(i))
                            sizeFound = CStr(sizes(j))
                            matchFound = True
                            Exit For
                        End If
                    Next j
                    If matchFound Then Exit For
                Next i
            End If
            
            ' Check for combined sizes
            If sizeFound = "" Then
                For i = LBound(combinedSizes) To UBound(combinedSizes)
                    If InStr(LCase(cleanedWord), LCase(combinedSizes(i))) > 0 Then
                        sizeFound = combinedSizes(i)
                        Exit For
                    End If
                Next i
            End If
            
            ' Check for numbers above 250 without sizes
            If IsNumeric(cleanedWord) Then
                On Error Resume Next
                If CLng(cleanedWord) > 250 Then
                    isNumberAbove250 = True
                    numberFound = cleanedWord
                End If
                On Error GoTo 0
            End If
            
            ' Exit the loop early if both number and size are found
            If numberFound <> "" And sizeFound <> "" Then
                Exit For
            End If
        Next word
        
        ' Debug information to check the values found
        Debug.Print "Cell: " & cell.Address & " Name: " & nameFound & " Number: " & numberFound & " Size: " & sizeFound
        
        ' If both a number and a size are found, place the name in column E and the number and size in column F
        If numberFound <> "" And sizeFound <> "" Then
            cell.Offset(0, 2).Value = nameFound
            cell.Offset(0, 3).Value = "Number: " & numberFound & ", Size: " & sizeFound
        ' If a number > 250 is found without a size, place the name and number in columns E and F
        ElseIf isNumberAbove250 Then
            cell.Offset(0, 2).Value = nameFound
            cell.Offset(0, 3).Value = "Number: " & numberFound
        End If

        ' If both a number and a size are found, check stock
        If numberFound <> "" And sizeFound <> "" Then
            ' Find the stock cell in the stock sheet
            Set stockRange = stockSheet.Range("B:B")
            Set stockCell = stockRange.Find(What:=numberFound, LookIn:=xlValues, LookAt:=xlWhole)
            
            If Not stockCell Is Nothing Then
                stockValue = stockCell.Offset(0, -1).Value ' Get the stock value from column P
                
                If stockValue > 0 Then
                    ' Decrease the stock value by 1
                    stockCell.Offset(0, -1).Value = stockValue - 1
                    
                    ' Place the data in columns G and H
                    ws.Cells(cell.Row, "G").Value = nameFound
                    ws.Cells(cell.Row, "H").Value = numberFound & " " & sizeFound
                Else
                    ' Place the data in columns R and S
                    ws.Cells(cell.Row, "R").Value = nameFound
                    ws.Cells(cell.Row, "S").Value = numberFound & " " & sizeFound
                End If
            End If
        ElseIf isNumberAbove250 Then
            ' Handle numbers above 250 without sizes
            Set stockRange = stockSheet.Range("B:B")
            Set stockCell = stockRange.Find(What:=numberFound, LookIn:=xlValues, LookAt:=xlWhole)
            
            If Not stockCell Is Nothing Then
                stockValue = stockCell.Offset(0, 15).Value ' Get the stock value from column P
                
                If stockValue > 0 Then
                    ' Decrease the stock value by 1
                    stockCell.Offset(0, 15).Value = stockValue - 1
                    
                    ' Place the data in columns G and H
                    ws.Cells(cell.Row, "G").Value = nameFound
                    ws.Cells(cell.Row, "H").Value = numberFound
                Else
                    ' Place the data in columns R and S
                    ws.Cells(cell.Row, "R").Value = nameFound
                    ws.Cells(cell.Row, "S").Value = numberFound
                End If
            Else
                Debug.Print "Number not found in stock: " & numberFound
            End If
        End If
    Next cell
End Sub

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by geniuszly** [comment]
Hi! It seems that the issue with your code might be related to how the logic is structured when checking the stock values. The logic should first check if the stock value is greater than 0 and, only then, copy it to the "shopping cart." If the stock value is 0 or less, it should be added to the "reserve list." However, if the code is adding items to the reserve list regardless of the stock value, it suggests that the check might not be implemented correctly.
You use the `stockCell.Offset(0, -1)` to access the stock value. If the stocks in your table are not in the column before the number, then the offset is incorrect.
You are using `stockSheet.Range("B:B")` to search for column B. This assumes that the product numbers are in column B. Make sure it matches your data structure. 
It is possible that the conditions for processing numbers greater than 250 overlap with the conditions for numbers and sizes, which may cause addition to the reserve list even if stocks are available.

Here is an example of how you can slightly adjust your code:
```vba
Sub FindNumbersAndSizes()
    Dim ws As Worksheet
    Dim rng As Range
    Dim cell As Range
    Dim numberFound As String
    Dim sizeFound As String
    Dim numbers As Variant
    Dim sizes As Variant
    Dim extraSizes As Variant
    Dim combinedSizes As Variant
    Dim i As Integer
    Dim j As Integer
    Dim nameFound As String
    Dim words As Variant
    Dim word As Variant
    Dim cleanedWord As String
    Dim combined As String
    Dim matchFound As Boolean
    Dim isNumberAbove250 As Boolean
    Dim stockSheet As Worksheet
    Dim stockRange As Range
    Dim stockCell As Range
    Dim stockValue As Long

    ' Set the worksheet (change "Liveshow" to the name of your sheet)
    Set ws = ThisWorkbook.Sheets("Liveshow")
    
    ' Set the stock sheet (change "StockSheet" to the name of your stock sheet)
    Set stockSheet = ThisWorkbook.Sheets("StockSheet")
    
    ' Set the range to the column you want to search (e.g., column C)
    Set rng = ws.Range("C1:C" & ws.Cells(ws.Rows.Count, "C").End(xlUp).Row)
    
    ' Define the numbers and sizes to search for
    numbers = Array(1 To 250) ' Adjusted array initialization for brevity

    ' Define size mapping
    sizes = Array("34", "xs", "extra small", _
                  "36", "s", "small", _
                  "38", "m", "medium", _
                  "40", "l", "large", _
                  "42", "xl", "extra large", _
                  "44", "xxl", "2xl", _
                  "46", "xxxl", "3xl")

    ' Define the combined sizes to search for
    combinedSizes = Array("s/m", "m/l", "l/xl")
    
    ' Define extra sizes to search for
    extraSizes = Array("85", "95", "105")
    
    ' Loop through each cell in the range
    For Each cell In rng
        numberFound = ""
        sizeFound = ""
        nameFound = ""
        matchFound = False
        isNumberAbove250 = False
        
        ' Get the name from column B
        nameFound = cell.Offset(0, -1).Value
        
        ' Split the cell value into words
        words = Split(cell.Value, " ")
        
        ' Check each word for numbers and sizes
        For Each word In words
            ' Remove periods and dashes from the word
            cleanedWord = Replace(word, ".", "")
            cleanedWord = Replace(cleanedWord, "-", "")
            
            ' Check for numbers
            For i = LBound(numbers) To UBound(numbers)
                If cleanedWord = CStr(numbers(i)) Then
                    numberFound = CStr(numbers(i))
                    Exit For
                End If
            Next i
            
            ' Check for sizes
            For i = LBound(sizes) To UBound(sizes) Step 3
                If LCase(cleanedWord) = LCase(CStr(sizes(i))) Or _
                   LCase(cleanedWord) = LCase(CStr(sizes(i + 1))) Or _
                   LCase(cleanedWord) = LCase(CStr(sizes(i + 2))) Then
                    sizeFound = CStr(sizes(i))
                    Exit For
                End If
            Next i
            
            ' Check for extra sizes
            If sizeFound = "" Then
                For i = LBound(extraSizes) To UBound(extraSizes)
                    If LCase(cleanedWord) = LCase(extraSizes(i)) Then
                        sizeFound = CStr(extraSizes(i))
                        Exit For
                    End If
                Next i
            End If
            
            ' Check for combined numbers and sizes
            If numberFound = "" Or sizeFound = "" Then
                For i = LBound(numbers) To UBound(numbers)
                    For j = LBound(sizes) To UBound(sizes)
                        combined = CStr(numbers(i)) & CStr(sizes(j))
                        If InStr(LCase(cleanedWord), LCase(combined)) > 0 Then
                            numberFound = CStr(numbers(i))
                            sizeFound = CStr(sizes(j))
                            matchFound = True
                            Exit For
                        End If
                    Next j
                    If matchFound Then Exit For
                Next i
            End If
            
            ' Check for combined sizes
            If sizeFound = "" Then
                For i = LBound(combinedSizes) To UBound(combinedSizes)
                    If InStr(LCase(cleanedWord), LCase(combinedSizes(i))) > 0 Then
                        sizeFound = combinedSizes(i)
                        Exit For
                    End If
                Next i
            End If
            
            ' Check for numbers above 250 without sizes
            If IsNumeric(cleanedWord) Then
                On Error Resume Next
                If CLng(cleanedWord) > 250 Then
                    isNumberAbove250 = True
                    numberFound = cleanedWord
                End If
                On Error GoTo 0
            End If
            
            ' Exit the loop early if both number and size are found
            If numberFound <> "" And sizeFound <> "" Then
                Exit For
            End If
        Next word
        
        ' Debug information to check the values found
        Debug.Print "Cell: " & cell.Address & " Name: " & nameFound & " Number: " & numberFound & " Size: " & sizeFound
        
        ' If both a number and a size are found, place the name in column E and the number and size in column F
        If numberFound <> "" And sizeFound <> "" Then
            cell.Offset(0, 2).Value = nameFound
            cell.Offset(0, 3).Value = "Number: " & numberFound & ", Size: " & sizeFound
        ' If a number > 250 is found without a size, place the name and number in columns E and F
        ElseIf isNumberAbove250 Then
            cell.Offset(0, 2).Value = nameFound
            cell.Offset(0, 3).Value = "Number: " & numberFound
        End If

        ' If both a number and a size are found, check stock
        If numberFound <> "" And sizeFound <> "" Then
            ' Find the stock cell in the stock sheet
            Set stockRange = stockSheet.Range("B:B")
            Set stockCell = stockRange.Find(What:=numberFound, LookIn:=xlValues, LookAt:=xlWhole)
            
            If Not stockCell Is Nothing Then
                stockValue = stockCell.Offset(0, -1).Value ' Get the stock value from column P
                
                If stockValue > 0 Then
                    ' Decrease the stock value by 1
                    stockCell.Offset(0, -1).Value = stockValue - 1
                    
                    ' Place the data in columns G and H
                    ws.Cells(cell.Row, "G").Value = nameFound
                    ws.Cells(cell.Row, "H").Value = numberFound & " " & sizeFound
                Else
                    ' Place the data in columns R and S
                    ws.Cells(cell.Row, "R").Value = nameFound
                    ws.Cells(cell.Row, "S").Value = numberFound & " " & sizeFound
                End If
            End If
        ElseIf isNumberAbove250 Then
            ' Handle numbers above 250 without sizes
            Set stockRange = stockSheet.Range("B:B")
            Set stockCell = stockRange.Find(What:=numberFound, LookIn:=xlValues, LookAt:=xlWhole)
            
            If Not stockCell Is Nothing Then
                stockValue = stockCell.Offset(0, 15).Value ' Get the stock value from column P
                
                If stockValue > 0 Then
                    ' Decrease the stock value by 1
                    stockCell.Offset(0, 15).Value = stockValue - 1
                    
                    ' Place the data in columns G and H
                    ws.Cells(cell.Row, "G").Value = nameFound
                    ws.Cells(cell.Row, "H").Value = numberFound
                Else
                    ' Place the data in columns R and S
                    ws.Cells(cell.Row, "R").Value = nameFound
                    ws.Cells(cell.Row, "S").Value = numberFound
                End If
            Else
                Debug.Print "Number not found in stock: " & numberFound
            End If
        End If
    Next cell
End Sub
```
